pkg/controller/volume/persistentvolume/pv_controller_base.go

```diff
@@ -244,20 +249,63 @@ func (ctrl *PersistentVolumeController) updateClaim(claim *v1.PersistentVolumeCl
 	}
 }
 
+// Unit test [5-5] [5-6] [5-7]
 // deleteClaim runs in worker thread and handles "claim deleted" event.
 func (ctrl *PersistentVolumeController) deleteClaim(claim *v1.PersistentVolumeClaim) {
 	_ = ctrl.claims.Delete(claim)
-	klog.V(4).Infof("claim %q deleted", claimToClaimKey(claim))
+	claimKey := claimToClaimKey(claim)
+	klog.V(4).Infof("claim %q deleted", claimKey)
+	// There are couple of scenarios here for metric recording:
+	// 1. Claim is NOT bound to any PV for some reason. In this case, clean provision operation
+	//    start timestamp from controller cache (operationTimestamps) if exists.
+	// 2. Claim is bound to a PV. In this case, there should NOT be provision operation start timestamp
+	//    cached in controller.
+	//    The bound PV will be recycled/retained/deleted based on corresponding retain policy
+	//    insert a deletion operation start timestamp into controller cache for the claim
+	//    if there is not one yet ONLY if the retain policy of PV is deletion
+	//    when a volume deleted event is captured, an end to end deletion latency will be reported
+	//    if error happened during "deleteVolumeOperation" call, error metric will be reported
+	// clean any possible unfinished provisioning start timestamp from cache
+	provisionTsKey := ctrl.createOperationTimestampCacheKey("provision", claimKey)
+
+	if _, exists := ctrl.operationTimestamps.Get(provisionTsKey); exists {
+		ctrl.operationTimestamps.Delete(provisionTsKey)
+	}
 
 	volumeName := claim.Spec.VolumeName
 	if volumeName == "" {
-		klog.V(5).Infof("deleteClaim[%q]: volume not bound", claimToClaimKey(claim))
+		klog.V(5).Infof("deleteClaim[%q]: volume not bound", claimKey)
 		return
 	}
+	// the claim is bound to a PV, insert start timestamp for deletion into controller
```

Comment:
I thought about moving this logic to pv_controller instead of putting in pv_controller_base which will indeed make the code cleaner. The issue is reclaimVolume could be called under different situations, for example.
1. reclaimVolume can be triggered by either delete a volume or delete a claim. We do not really want to record the metric for the first case.
2. reclaimVolume can also be triggered by situations like two volumes bound to a same claim, one of them will be reclaimed, in this case we also do not want to record end to end metric
Putting a start time stamp to when a claim is deleted from system + the claim is bound to a PV seems to be an appropriate place.
Any other suggestions?

Reply:
For 1, I couldn't find the logic where reclaimVolume is called by deleting a volume.  I could only find the path where it gets called by deleting the claim.

For 2, I think this is a valid scenario to measure.  It's not user-triggered, however, it is something that we still have to call the driver to do work.

Reply:
done